### ISSUE TITLE
Certificate Pinning - Not Found

### ISSUE THREAD

**Issue by Ac3n34d-X** [shroudedcode/apk-mitm#8]
So This App Actually Has Certificate Pinning But When You Run it Through APK-MITM it says no certificate Pinning Method Found
When You Try to Debug the App It Sends Requests to An APi on port 433 
https://www.sendspace.com/file/q3v2c2

**Comment by shroudedcode** [comment]
> When You Run it Through APK-MITM it says no certificate Pinning Method Found

There's multiple ways apps can implement certificate pinning. `apk-mitm` helps with the following two:

- Using `X509TrustManager`
- Using [the `<pin-set>` option](https://developer.android.com/training/articles/security-config#pin-set) in the Network Security Configuration

The message you got (I assume it was *"No certificate pinning logic found."*) means that `apk-mitm` wasn't able to find an implementation of `X509TrustManager` in the app's code. The app's developers might have implemented certificate pinning in a different way I'm not aware of or maybe they're not using certificate pinning at all. Why are you so sure they are?

**Comment by Ac3n34d-X** [comment]
Because if you test the app straight without fiddler or any mitm tool, it works fine

**Comment by shroudedcode** [comment]
I just looked at the app's code and realized that the app does in fact use `X509TrustManager`, but all the certificate pinning methods use the `final` keyword, for example here:

```smali
.method public final checkClientTrusted([Ljava/security/cert/X509Certificate;Ljava/lang/String;)V
```

`apk-mitm` currently only looks for [method signatures without `final`](https://github.com/shroudedcode/apk-mitm/blob/31567688caf652b3eb6647fc00812330ff1ae5ba/src/tasks/disable-certificate-pinning.ts#L10-L12) because all the apps I've looked at so far don't use `final` for these methods. Since it's perfectly reasonable to use `final` here and there are probably more apps doing this I see this as a bug in `apk-mitm`.

Thanks for bringing this to my attention! I'll let you know when I've released a fix.
